OpenKeychain/src/main/java/org/sufficientlysecure/keychain/provider/ProviderWriter.java

```diff
@@ -0,0 +1,1203 @@
+        mProviderHelper = helper;
+        mContentResolver = resolver;
+        mContext = mProviderHelper.mContext;
+    }
+
+    public static ProviderWriter newInstance(ProviderHelper helper, ContentResolver resolver) {
+        return new ProviderWriter(helper, resolver);
+    }
+
+    // bits, in order: CESA. make SURE these are correct, we will get bad log entries otherwise!!
+    static final LogType LOG_TYPES_FLAG_MASTER[] = new LogType[]{
+            LogType.MSG_IP_MASTER_FLAGS_XXXX, LogType.MSG_IP_MASTER_FLAGS_CXXX,
+            LogType.MSG_IP_MASTER_FLAGS_XEXX, LogType.MSG_IP_MASTER_FLAGS_CEXX,
+            LogType.MSG_IP_MASTER_FLAGS_XXSX, LogType.MSG_IP_MASTER_FLAGS_CXSX,
+            LogType.MSG_IP_MASTER_FLAGS_XESX, LogType.MSG_IP_MASTER_FLAGS_CESX,
+            LogType.MSG_IP_MASTER_FLAGS_XXXA, LogType.MSG_IP_MASTER_FLAGS_CXXA,
+            LogType.MSG_IP_MASTER_FLAGS_XEXA, LogType.MSG_IP_MASTER_FLAGS_CEXA,
+            LogType.MSG_IP_MASTER_FLAGS_XXSA, LogType.MSG_IP_MASTER_FLAGS_CXSA,
+            LogType.MSG_IP_MASTER_FLAGS_XESA, LogType.MSG_IP_MASTER_FLAGS_CESA
+    };
+
+    // same as above, but for subkeys
+    static final LogType LOG_TYPES_FLAG_SUBKEY[] = new LogType[]{
+            LogType.MSG_IP_SUBKEY_FLAGS_XXXX, LogType.MSG_IP_SUBKEY_FLAGS_CXXX,
+            LogType.MSG_IP_SUBKEY_FLAGS_XEXX, LogType.MSG_IP_SUBKEY_FLAGS_CEXX,
+            LogType.MSG_IP_SUBKEY_FLAGS_XXSX, LogType.MSG_IP_SUBKEY_FLAGS_CXSX,
+            LogType.MSG_IP_SUBKEY_FLAGS_XESX, LogType.MSG_IP_SUBKEY_FLAGS_CESX,
+            LogType.MSG_IP_SUBKEY_FLAGS_XXXA, LogType.MSG_IP_SUBKEY_FLAGS_CXXA,
+            LogType.MSG_IP_SUBKEY_FLAGS_XEXA, LogType.MSG_IP_SUBKEY_FLAGS_CEXA,
+            LogType.MSG_IP_SUBKEY_FLAGS_XXSA, LogType.MSG_IP_SUBKEY_FLAGS_CXSA,
+            LogType.MSG_IP_SUBKEY_FLAGS_XESA, LogType.MSG_IP_SUBKEY_FLAGS_CESA
+    };
+
+    /**
+     * Saves an UncachedKeyRing of the public variant into the db.
+     * <p/>
+     * This method will not delete all previous data for this masterKeyId from the database prior
+     * to inserting. All public data is effectively re-inserted, secret keyrings are left deleted
+     * and need to be saved externally to be preserved past the operation.
+     */
+    @SuppressWarnings("unchecked")
+    private int saveCanonicalizedPublicKeyRing(CanonicalizedPublicKeyRing keyRing,
+                                               Progressable progress, boolean selfCertsAreTrusted) {
+        selfCertsAreTrusted = selfCertsAreTrusted || mProviderHelper.mConsolidatingOwnPublic;
+
+        // start with ok result
+        int result = SaveKeyringResult.SAVED_PUBLIC;
+
+        long masterKeyId = keyRing.getMasterKeyId();
+        UncachedPublicKey masterKey = keyRing.getPublicKey();
+
+        ArrayList<ContentProviderOperation> operations;
+        try {
+
+            mProviderHelper.log(LogType.MSG_IP_PREPARE);
+            result += 1;
```

Comment:
I think this used to be ```mIndent+=1``` before it was copied over. This also highlights an issue with logging in `ProviderWriter` and `ProviderReader` in general - I'm not very sure whether accessing the ```ProviderHelper```'s operation log this way is ideal (getting it to work would involve changing `ProviderHelper.mIndent` too).